Fix thread static cleanup paths

davidwrighton opened the pull request:
- Do not destroy GC handles while holding the spin lock
- Free the pLoaderHandle array when the thread is terminated

Fixes #104449

davidwrighton committed:
Fix thread static cleanup paths
- Do not destroy GC handles while holding the spin lock
- Free the pLoaderHandle array when the thread is terminated

davidwrighton committed:
Fix assert

jkotas commented on src/coreclr/vm/threadstatics.cpp:
```diff
@@ -405,6 +405,13 @@ void FreeLoaderAllocatorHandlesForTLSData(Thread *pThread)
         }
+
+        pThread->cLoaderHandles = -1; // Sentinel value indicating that there are no LoaderHandles and the thread is permanently dead.
+        if (pThread->pLoaderHandles != NULL)
```
Nit: This should be always true since `cLoaderHandles > 0` here. Also, the `delete` can be called outside the lock.

jkotas commented on src/coreclr/vm/threadstatics.cpp:
```diff
@@ -405,6 +405,13 @@ void FreeLoaderAllocatorHandlesForTLSData(Thread *pThread)
             }
         }
+
+        pThread->cLoaderHandles = -1; // Sentinel value indicating that there are no LoaderHandles and the thread is permanently dead.
```
Set the sentinel before we start cleaning the entries?

AaronRobinsonMSFT commented on src/coreclr/vm/threadstatics.cpp:
```diff
@@ -431,34 +438,49 @@ void FreeThreadStaticData(Thread* pThread)
+        pOldCollectibleTlsArrayData = pThreadLocalData->pCollectibleTlsArrayData;
+        oldCollectibleTlsDataCount = pThreadLocalData->cCollectibleTlsData;
+
+        pThreadLocalData->pCollectibleTlsArrayData = 0;
+        pThreadLocalData->cCollectibleTlsData = 0;
+        pThreadLocalData->pNonCollectibleTlsArrayData = 0;
+        pThreadLocalData->cNonCollectibleTlsData = 0;
```
```suggestion
        pThreadLocalData->pCollectibleTlsArrayData = NULL;
        pThreadLocalData->cCollectibleTlsData = 0;
        pThreadLocalData->pNonCollectibleTlsArrayData = NULL;
        pThreadLocalData->cNonCollectibleTlsData = 0;
```

AaronRobinsonMSFT commented on src/coreclr/vm/threadstatics.cpp:
```diff
@@ -431,34 +438,49 @@ void FreeThreadStaticData(Thread* pThread)
-    pThreadLocalData->cCollectibleTlsData = 0;
-    pThreadLocalData->pNonCollectibleTlsArrayData = 0;
-    pThreadLocalData->cNonCollectibleTlsData = 0;
+    if (pOldCollectibleTlsArrayData != NULL)
+    {
+        delete[] (uint8_t*)pOldCollectibleTlsArrayData;
+    }
```
```suggestion
    delete[] (uint8_t*)pOldCollectibleTlsArrayData;
```

davidwrighton committed:
When using a ThreadStatics stress test on collectible assemblies, a few more issues were found
- Fix issue where the LoaderAllocator's SegmentedHandleIndex wasn't being freed
- Fix issue where the logic to re-use TLSIndex values wasn't working properly

davidwrighton commented:
/backport to release/9.0
